File: src/benchmark/bench3.nim
import ../oche

type
  Vec3 {.oche.} = object
    x, y, z: float

# ===============================================
# SCENARIO 1: Large Array Generation & Summation
# ===============================================

# 1.1 Copy Mode (Old FFI Way - High Overhead)
proc generateLargeArrayCopy(n: int): seq[Vec3] {.oche.} =
  result = newSeq[Vec3](n)
  for i in 0..<n:
    result[i] = Vec3(x: i.float, y: (i*2).float, z: (i*3).float)

# 1.2 View Mode (Zero-Copy Read-Only)
var cachedArray: seq[Vec3]
proc generateLargeArrayView(n: int): seq[Vec3] {.oche: view.} =
  cachedArray = newSeq[Vec3](n)
  for i in 0..<n:
    cachedArray[i] = Vec3(x: i.float, y: (i*2).float, z: (i*3).float)
  return cachedArray

# ===============================================
# SCENARIO 2: Particle System Update (Mutation)
# ===============================================

var particleBuffer: OcheBuffer[Vec3]

proc initParticles(n: int): OcheBuffer[Vec3] {.oche.} =
  particleBuffer = newOche[Vec3](n)
  for i in 0..<n:
    particleBuffer[i] = Vec3(x: 0, y: 0, z: 0)
  return particleBuffer

proc nimUpdateParticles(dt: float) {.oche.} =
  for i in 0..<particleBuffer.len:
    particleBuffer[i].x += dt
    particleBuffer[i].y += dt * 2.0
    particleBuffer[i].z += dt * 3.0

generate("nlib_bench3.dart")

File: src/benchmark/bench3.dart
import 'nlib_bench3.dart';

void runBench(String name, Function testCase) {
  print("\n[*] BENCHMARK: $name");
  final sw = Stopwatch()..start();
  testCase();
  sw.stop();
  print("  - Time: ${sw.elapsedMilliseconds}ms");
}

void main() {
  print("==========================================");
  print("     🏃‍♂️ OCHE ZERO-COPY BENCHMARK 🏃‍♂️    ");
  print("==========================================\n");

  const SIZE = 10000000; // 10 Million Items

  print(">>> SCENARIO 1: Generate & Sum 10M Objects <<<\n");

  // 1. Pure Dart creation and summation
  runBench("Pure Dart (Object Creation & Sum)", () {
    final list = List.generate(
      SIZE,
      (i) => Vec3(x: i.toDouble(), y: i * 2.0, z: i * 3.0),
      growable: false,
    );
    double sum = 0;
    for (var i = 0; i < list.length; i++) {
      sum += list[i].x + list[i].y + list[i].z;
    }
  });

  // 2. Nim -> Dart (Copy Mode)
  runBench("Oche Copy Mode (Old FFI Way / Deep Copy)", () {
    final list = oche.generateLargeArrayCopy(SIZE);
    double sum = 0;
    for (var i = 0; i < list.length; i++) {
      sum += list[i].x + list[i].y + list[i].z;
    }
  });

  // 3. Nim -> Dart (Zero-Copy View Mode)
  runBench("Oche Zero-Copy View Mode (The New Way!)", () {
    final view = oche.generateLargeArrayView(SIZE);
    double sum = 0;
    for (var i = 0; i < view.length; i++) {
      sum += view[i].x + view[i].y + view[i].z;
    }
  });

  print("\n>>> SCENARIO 2: Particle Update (Mutation of 10M Vec3) <<<\n");

  // Pure Dart updating particles
  runBench("Pure Dart Mutable", () {
    final particles = List.generate(
      SIZE,
      (i) => Vec3(x: 0, y: 0, z: 0),
      growable: false,
    );
    for (var i = 0; i < particles.length; i++) {
      particles[i] = Vec3(
        x: particles[i].x + 0.16,
        y: particles[i].y + 0.32,
        z: particles[i].z + 0.48,
      );
    }
  });

  // Oche Native Code Memory Shared Update
  runBench("Oche Shared Buffer (Nim Updates Memory in C)", () {
    final shared = oche.initParticles(SIZE);
    oche.nimUpdateParticles(0.16); // Direct RAM update
  });

  print("\n==========================================");
  print("   🏁 BENCHMARK COMPLETE! 🏁      ");
  print("==========================================");
}
